server/src/main/java/io/crate/expression/scalar/regex/RegexpMatchScalar.java

```diff
@@ -0,0 +1,114 @@
+import io.crate.expression.operator.Operator;
+import io.crate.expression.scalar.ScalarFunctionModule;
+import io.crate.expression.symbol.Literal;
+import io.crate.lucene.match.CrateRegexQuery;
+import io.crate.metadata.NodeContext;
+import io.crate.metadata.Reference;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.BoundSignature;
+import io.crate.metadata.functions.Signature;
+import io.crate.types.DataTypes;
+
+public class RegexpMatchScalar extends Scalar<String[], String> {
+    public static final String NAME = "regexp_match_scalar";
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(
+            Signature.scalar(
+                NAME,
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.STRING.getTypeSignature(),
+                Operator.RETURN_TYPE.getTypeSignature()
+            ),
+            RegexpMatchScalar::new
+        );
+
+        module.register(
+            Signature.scalar(
+                NAME,
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.STRING.getTypeSignature(),
+                DataTypes.STRING.getTypeSignature(),
+                Operator.RETURN_TYPE.getTypeSignature()
+            ),
+            RegexpMatchScalar::new
+        );
+    }
+
+    public RegexpMatchScalar(Signature signature, BoundSignature boundSignature) {
+        super(signature, boundSignature);
+    }
+
+    @Override
+    public String[] evaluate(TransactionContext txnCtx, NodeContext nodeCtx, Input<String>[] args) {
+        assert args.length == 2 || args.length == 3 : "number of args must be 2 or 3";
+        String source = args[0].value();
+        if (source == null) {
+            return null;
+        }
+
+        String pattern = args[1].value();
+        if (pattern == null) {
+            return null;
+        }
+
+        String flags = null;
+
+        if (args.length == 3) {
+            flags = args[2].value();
+            if (!isGlobal(flags)) {
+                pattern = Pattern.compile(pattern, parseFlags(flags)).pattern();
+            }
+        }
+
+        List<String> matches = new ArrayList<>();
+        java.util.regex.Matcher matcher = java.util.regex.Pattern.compile(pattern).matcher(source);
+        while (matcher.find()) {
+            matches.add(matcher.group());
+        }
+        return matches.toArray(new String[0]);
+    }
+
+    @Override
+    public Query toQuery(Reference ref, Literal<?> literal) {
+        String pattern = (String) literal.value();
+        int flags = 0;
+
+        String flagsStr = null;
+        int flagSeparatorIndex = pattern.lastIndexOf('/');
+        if (flagSeparatorIndex != -1 && flagSeparatorIndex < pattern.length() - 1) {
+            pattern = pattern.substring(0, flagSeparatorIndex);
+            flagsStr = pattern.substring(flagSeparatorIndex + 1);
+            flags = parseFlags(flagsStr);
+        }
```

Comment:
Any optional flag would be the 3rd argument and not part of the pattern, see https://www.postgresql.org/docs/current/functions-matching.html.
To support this, a custom `toQuery(Function function, LuceneQueryBuilder.Context context)` would be needed to take the 3rd argument into account.